packages/vm/src/testing/mock.rs

```diff
@@ -134,6 +136,25 @@ impl Api for MockApi {
         };
         (result, gas_info)
     }
+
+    fn secp256k1_verify(
+        &self,
+        message_hash: &[u8],
+        signature: &[u8],
+        public_key: &[u8],
+    ) -> BackendResult<()> {
+        let gas_info = GasInfo::with_cost(GAS_COST_VERIFY_SIGNATURE);
+
+        if let Some(backend_error) = self.backend_error {
+            return (Err(BackendError::unknown(backend_error)), gas_info);
+        }
+
+        let result = match crypto::secp256k1_verify(message_hash, signature, public_key) {
+            Ok(_) => Ok(()),
+            Err(err) => Err(BackendError::crypto_err(err.to_string())),
+        };
+        (result, gas_info)
+    }
```

Comment:
This mock needs to be moved into the MockApi from cosmwasm-std, which is a different API

Reply:
AFAIK, that implies the whole functionality should be moved to (or rewritted in) `std`.

`cosmwasm-vm` already has a dependency on `cosmwasm-std` (because of the `iterator` and `ibc`-related features, I think). So, I cannot introduce a (cyclical) dependency on `cosmwasm-vm`.

Reply:
If I'm not mistaken, there are basically three or four options around this:

1. Rewrite the functionality, just for `cosmwasm-std::MockApi`. This is what is being done for `canonical`/`human_address`. This is not a good idea: it leads to brittleness, and kicks the can of the maintenance burden down the road. It also adds / requires dependencies in both packages, etc.
2. Move the `crypto` stuff to `std`, use it in `MockApi`, and import it in `vm`. Just to be accessed from `std` through `ExternalApi`. This is not very nice, but definitively better than 1., and, despite its quirkyness, generally not a bad thing to do IMO.
3. Invert the dependency relationship between `vm` and `std`. Moving stuff around, etc. This can prove to be more difficult than it looks, and maybe not the best idea.
4. Add a new `commons`, or `shared`, or whatever, crate / package. This looks overkill.

So, I would go for 2. (easily doable), and in any case 3. (more work / changes, but (perhaps?) better dependency graph in the end).

What do you think?

Reply:
Or would you'd be happy / it will be enough with a mock verify function that return always true?

Reply:
OK, implemented that in the latest commits, and I'll change it if needed.

Reply:
I think it is important to make verification success/failures available to unit tests to encourage testing both code paths. I dee the problem. I think the cleanest way is to introduce an internal `cosmwasm-crypto`, that is shared between `cosmwasm_std::testing` (i.e. is never compiled into the real contract) and `cosmwasm_vm`.

Reply:
I think the same. FYI, I did the cross-packages impl (point 2. above), just to check how it was. And, it doesn't look so bad. I even managed to avoid compiling it for wasm in `cosmwasm-std` (by using `#cfg(not(target_arch = "wasm32"))`).

I think that an internal crate is better, though. Let's stabilize the definitions here, and then we can merge, or I can branch and refactor in a different PR.

Reply:
> FYI, I did the cross-packages impl (point 2. above), just to check how it was. And, it doesn't look so bad. I even managed to avoid compiling it for wasm (by using #cfg(not(target_arch = "wasm32"))).

In this PR or locally? Ideally we remove the whole module `cosmwasm_std::testing` from Wasm builds to prevent misusage. But then again, maybe users want to run unit tests in Wasm instead of their dev environment (just like https://github.com/CosmWasm/cosmwasm/issues/348 and https://github.com/CosmWasm/cosmwasm/issues/524).

Reply:
In a different branch, locally. I can publish it if you like.

Reply:
I'd just pull out the new crate instead. But we can use your version as an intermediate step if you think it helps.

Reply:
Take a look at https://github.com/CosmWasm/cosmwasm/tree/crypto-api-rust-only-std. It doesn't have all the recent changes but, you get the idea.

Reply:
> I'd just pull out the new crate instead. But we can use your version as an intermediate step if you think it helps.

OK with me.